src/rcc.rs

```diff
@@ -1,26 +1,341 @@
+            };
+            MediumSpeedInternalRC {
+                f: f,
+                auto_cal: a,
+                bits: bits,
+            }
+        }
+
+        pub fn bits(&self) -> u8 {
+            self.bits
+        }
+
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            rcc.cr.modify(|_, w| unsafe {
+                w.msirange()
+                    .bits(self.bits())
+                    .msirgsel()
+                    .set_bit()
+            });
+            while rcc.cr.read().msirdy().bit_is_clear() {}
+
+            if self.auto_cal {
+                // FIXME This... may not work? I'm not sure if I've got a board problem or using
+                // the LSE requires some precondition I'm missing. In either case, LSERDY is never
+                // set by the hardware, so auto_cal doesn't succeed.
+                rcc.apb1enr1.modify(|_,w| w.pwren().set_bit());
+
+                rcc.bdcr.modify(|_,w| w.lseon().clear_bit());
+                while rcc.bdcr.read().lserdy().bit_is_set() {}
+                rcc.bdcr.modify(|_,w| unsafe {
+                    w.lsedrv()
+                        .bits(0b11)
+                        .lseon()
+                        .set_bit()
+                });
+                while rcc.bdcr.read().lserdy().bit_is_clear() {}
+                rcc.cr.modify(|_,w| w.msipllen().set_bit());
+            }
+            (self.freq(), 0b00)
+        }
+    }
+
+    impl InputClock for MediumSpeedInternalRC {
+        fn freq(&self) -> u32 {
+            self.f
+        }
+    }
+
+    /// High-speed external 4-48 MHz oscillator
+    #[derive(Clone, Copy)]
+    pub struct HighSpeedExternalOSC(pub u32);
+    impl InputClock for HighSpeedExternalOSC {
+        fn freq(&self) -> u32 {
+            self.0
+        }
+    }
+
+    impl HighSpeedExternalOSC {
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            rcc.cr.modify(|_, w| w
+                          .hseon()
+                          .set_bit()
+                         );
+            while rcc.cr.read().hserdy().bit_is_clear() {}
+            (self.freq(), 0b10)
+        }
+    }
+
+    pub enum RtcClkSource {
+        None,
+        /// Internal 32 kHz RC
+        LSI,
+        /// External 32.768 kHz oscillator
+        LSE,
+        HSEDiv32,
+    }
+    impl RtcClkSource {
+        pub fn freq(&self, hse: Option<HighSpeedExternalOSC>) -> Result<u32, ()> {
+            match *self {
+                RtcClkSource::None => Err(()),
+                RtcClkSource::LSI => Ok(32_000),
+                RtcClkSource::LSE => Ok(32_768),
+                RtcClkSource::HSEDiv32 => {
+                    if let Some(clk) = hse {
+                        Ok(clk.freq() / 32)
+                    } else {
+                        Err(())
+                    }
+                },
+            }
+        }
+
+        pub fn bits(&self) -> u8 {
```

Comment:
Maybe use `repr(u8)` as I originally had with C style enum?
This way you can just make `self as u8`?